pkg/drivers/vfkit/vfkit.go

```diff
@@ -0,0 +1,513 @@
+		return state.Error, err
+	}
+	pid, err := strconv.Atoi(strings.TrimSpace(string(p)))
+	if err != nil {
+		return state.Error, err
+	}
+	if err := checkPid(pid); err != nil {
+		// No pid, remove pidfile
+		os.Remove(d.pidfilePath())
+		return state.Stopped, nil
+	}
+	ret, err := d.GetVFKitState()
+	if err != nil {
+		return state.Error, err
+	}
+	switch ret {
+	case "running", "VirtualMachineStateRunning":
+		return state.Running, nil
+	case "stopped", "VirtualMachineStateStopped":
+		return state.Stopped, nil
+	}
+	return state.None, nil
+}
+
+func (d *Driver) Create() error {
+	var err error
+	if d.SSHPort, err = d.GetSSHPort(); err != nil {
+		return err
+	}
+	b2dutils := mcnutils.NewB2dUtils(d.StorePath)
+	if err := b2dutils.CopyIsoToMachineDir(d.Boot2DockerURL, d.MachineName); err != nil {
+		return err
+	}
+	isoPath := d.ResolveStorePath(isoFilename)
+
+	log.Info("Extracting Kernel...")
+	if err := d.extractKernel(isoPath); err != nil {
+		return err
+	}
+
+	log.Info("Creating SSH key...")
+	if err := ssh.GenerateSSHKey(d.sshKeyPath()); err != nil {
+		return err
+	}
+
+	log.Info("Creating Disk image...")
+	if err := d.generateDiskImage(d.DiskSize); err != nil {
+		return err
+	}
+
+	if d.ExtraDisks > 0 {
+		log.Info("Creating extra disk images...")
+		for i := 0; i < d.ExtraDisks; i++ {
+			path := pkgdrivers.ExtraDiskPath(d.BaseDriver, i)
+			if err := pkgdrivers.CreateRawDisk(path, d.DiskSize); err != nil {
+				return err
+			}
+		}
+	}
+
+	log.Info("Starting vfkit VM...")
+	return d.Start()
+}
+
+func (d *Driver) Start() error {
+	machineDir := filepath.Join(d.StorePath, "machines", d.GetMachineName())
+
+	var startCmd []string
+
+	startCmd = append(startCmd,
+		"--memory", fmt.Sprintf("%d", d.Memory),
+		"--cpus", fmt.Sprintf("%d", d.CPU),
+		"--restful-uri", fmt.Sprintf("unix://%s", d.sockfilePath()))
+	var isoPath = filepath.Join(machineDir, isoFilename)
+	startCmd = append(startCmd,
+		"--device", fmt.Sprintf("virtio-blk,path=%s", isoPath))
+
+	var mac = d.MACAddress
+	startCmd = append(startCmd,
+		"--device", fmt.Sprintf("virtio-net,nat,mac=%s", mac))
+
+	startCmd = append(startCmd,
+		"--device", "virtio-rng")
+
+	startCmd = append(startCmd,
+		"--kernel", d.ResolveStorePath("bzimage"))
+	startCmd = append(startCmd,
+		"--kernel-cmdline", d.Cmdline)
+	startCmd = append(startCmd,
+		"--initrd", d.ResolveStorePath("initrd"))
+
+	for i := 0; i < d.ExtraDisks; i++ {
+		startCmd = append(startCmd,
+			"--device", fmt.Sprintf("virtio-blk,path=%s", pkgdrivers.ExtraDiskPath(d.BaseDriver, i)))
+	}
+
+	startCmd = append(startCmd,
+		"--device", fmt.Sprintf("virtio-blk,path=%s", d.diskPath()))
+
+	log.Debugf("executing: vfkit %s", strings.Join(startCmd, " "))
+	os.Remove(d.sockfilePath())
+	cmd := exec.Command("vfkit", startCmd...)
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+	pid := cmd.Process.Pid
+	if err := os.WriteFile(d.pidfilePath(), []byte(fmt.Sprintf("%v", pid)), 0600); err != nil {
+		return err
+	}
```

Comment:
Creating the pid file for vfkit is a bit racy, if we are killed after starting vfkit but before we create the pid file, we will never terminate vfkit.

I think we need to move this vfkit - it should provide a --pidfile option. It goes with the --daemonize option we also need.

Reply:
Another race which can happen is that `vfkit` can exit after `cmd.Start()` succeeds, for example if Apple's virtualization framework fails to start the VM and returns an error. If this happens, you'd have a pid file but no corresponding process.
I added this code in vfkit's test suite to catch this https://github.com/crc-org/vfkit/commit/a38d84c167f97a520f859e17283c40c299e68e03
It probably would be useful to have a way to tell vfkit "do not exit by yourself, I'll explicitly stop/kill/.. you when I no longer need your process to be around"